### ISSUE TITLE
[xstate-react] Selecting state.children with useSelector causes re-renders

### ISSUE THREAD

**Issue by fraserdrops** [statelyai/xstate#1967]
**Description**
Selecting state.children with useSelector causes re-renders any time an event happens.
e.g.
```
const selectChildren = (state) => state.children;
// any event sent to the service will cause the reference to children to change
const children = useSelector(service, selectChildren);
```
**Expected Result**
state.children is the same object unless an action causes a change to children (e.g. spawn)

**Actual Result**
state.children always references a different object.

**Reproduction**

- [Codesandbox (JavaScript)](https://codesandbox.io/s/xstate-select-children-rerender-fptuk)

**Comment by davidkpiano** [comment]
This optimization would have to be made in XState, as currently the `state.children` object is "rebuilt" for every state change.

We can do that, but the `useSelector` hook is still working as designed, and you need to take into special consideration selected non-primitive values, _especially_ from 3rd-party sources, where you cannot guarantee that they will have stable references.

What is your use-case for selecting `state.children`?

**Comment by FMcIntosh** [comment]
You're right, useSelector is working, but I found it surprising that `state.children` isn't stable. My initial conception was that only if the children were changed by an action would the reference change.  However, as you say XState is doing something else.

When playing around, `state.context` is stable (the top level object) and `state.context[x]` is stable (assuming `x` isn't changed from a call to `assign`. Almost every other object (state.meta, state.event, state.activities) is "rebuilt" on every state change. Is the guideline that only state.context should be treated as stable? Or is that also implementation detail?

It's worth noting that a workaround is to do a shallow compare `const children = useSelector(service, selectChildren, shallowCompare);`. If that's the recommended approach, I'd be happy to add something to the docs to note this.

My usecase is to pass services down to child components:
```
const selectChildren = (state) => state.children;
function App() {
  const service = useInterpret(AppMachine);
  const children = useSelector(service,  selectChildren);

  return (
    <div>
        <Sidebar service={children.sidebar} />
        <Grid service={children.grid} />
    </div>
  );
}
```
Another workaround would be to select each child individually:
```
const sidebar = useSelector(service,  selectSidebar);
const grid = useSelector(service,  selectGrid);
```
